### ISSUE TITLE
[proposal] DS allow more Activation-Injections

### ISSUE THREAD

**Issue by stbischof** [osgi/osgi#290]
The Metatype-,Declarative Services-,Converter-Spec work really great together. Especially the usage of Annotations with Meta-type-Annotations and the injection of the converted maps into the activation methods is really great!

But in some cases Annotations does not fit best. Especially if you do not need the ComponentPropertyType but more then one slightly different Configurations, the extension mechanism of Interfaces would be more useful.  Also for testing and when activating a Components in a non OSGI-World Interfaces would be nice.

So in wrote down the basic concepts behind the "workflow" and the Entity that are able to hold the config.

|                              | Annotations | Interface | Records | DTO |
|------------------------------|-------------|-----------|---------|-----|
| generate Metatype xml        | x           | x         | w       | ?   |
| generate Component xml       | x           | (x)*1       | w       | ?   |
| convert from Map             | x           | x         | (x)*3     | x   |
| inject on activation         | x           | (x)*2       | w       | ?   |
| use as ComponentPropertyType | x           | !         | !       | !   |

x - Spec
w - would be great to have
? - not sure
! - out of scope
*1 - bnd `-dsannotations-options: felixExtensions`
*2 - [felic-scr](https://github.com/apache/felix-dev/blob/3e5671ae7e5107f4f849ef9d5f0a89b1ba9d7439/scr/src/main/java/org/apache/felix/scr/impl/xml/XmlHandler.java#L215) `ATTR_CONFIGURE_WITH_INTERFACES`
*3 - [proofe of concept](https://github.com/stbischof/osgi/tree/Conv_Record)

In conclusion it would be greate to have Meta-type-Annotation  and DS-Injection Support for Interfaces and Records.

**Comment by bjhargrave** [comment]
One solution would be to support annotating the argument with the name of a user-supplied class which can coerce the map into the injected object type. Like Junit5 does.

**Comment by stbischof** [comment]
Interfaces could be analyzes do create MetatypeInformation (bnd) and inject into the service (felix)

What was the reason to not put this into the spec?

1 - bnd -dsannotations-options: felixExtensions
2 - [felic-scr](https://github.com/apache/felix-dev/blob/3e5671ae7e5107f4f849ef9d5f0a89b1ba9d7439/scr/src/main/java/org/apache/felix/scr/impl/xml/XmlHandler.java#L215) ATTR_CONFIGURE_WITH_INTERFACES
